File: FeatureEngineering/run_feature_engineering_smoke.py
from __future__ import annotations

import csv
import json
import os
import platform
import sys
import time
from datetime import datetime, timedelta
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
OUT_DIR = Path(os.environ.get("FEATURE_ENGINEERING_BENCHMARK_OUTPUT", ROOT / "FeatureEngineering" / "outputs" / datetime.now().strftime("%Y%m%d_%H%M%S")))
OUT_DIR.mkdir(parents=True, exist_ok=True)

POLARSFE_ROOT = ROOT.parent / "polars_feature_engineering"
if str(POLARSFE_ROOT) not in sys.path:
    sys.path.insert(0, str(POLARSFE_ROOT))

rows = []


def record(engine, family, n_rows, fn):
    start = time.perf_counter()
    status = "success"
    error = ""
    try:
        fn()
    except Exception as exc:  # noqa: BLE001
        status = "error"
        error = str(exc)
    elapsed = time.perf_counter() - start
    rows.append(
        {
            "engine": engine,
            "family": family,
            "rows": n_rows,
            "elapsed_seconds": round(elapsed, 6),
            "status": status,
            "error": error,
        }
    )


def main():
    import polars as pl

    n = 10000
    base_date = datetime(2024, 1, 1)
    df = pl.DataFrame(
        {
            "id": list(range(1, n + 1)),
            "x": [(i % 101) + 0.1 for i in range(n)],
            "y": [50 + ((i * 17) % 31) for i in range(n)],
            "cat": [["A", "B", "C", "D", "E"][i % 5] for i in range(n)],
            "cat2": [["K", "L", "M"][i % 3] for i in range(n)],
            "target": [["no", "yes"][i % 2] for i in range(n)],
            "group": [f"g{(i % 200) + 1}" for i in range(n)],
            "date": [base_date + timedelta(days=i % 365) for i in range(n)],
            "text": [["Hello WORLD", "two words", "ABC123!", "small", "", None][i % 6] for i in range(n)],
        }
    )

    import PolarsFE

    plan = PolarsFE.polars_feature_plan(
        numeric={"columns": ["x", "y"], "transforms": ["log1p", "sqrt", "standardize", "winsorize"], "winsorize_probs": [0.01, 0.99]},
        categorical={"columns": ["cat"], "top_n": 4, "rare_level": "__RARE__", "unseen_level": "__UNSEEN__", "one_hot": True, "keep_original": True},
        calendar={"columns": ["date"], "features": ["year", "month", "wday", "quarter", "is_weekend"]},
        text={"columns": ["text"], "features": ["char_count", "word_count", "digit_count", "blank"]},
        missingness={"columns": ["x", "cat", "text"], "suffix": "_is_missing"},
        interactions={
            "numeric_pairs": [["x", "y"]],
            "categorical_numeric": [{"categorical": "cat", "numeric": "x"}],
            "categorical_pairs": [["cat", "cat2"]],
            "max_features": 20,
        },
    )

    record("PolarsFE vNext", "combined_plan", n, lambda: PolarsFE.polars_fit_transform_feature_plan(df, plan))

    partition_plan = PolarsFE.polars_partition_plan(
        method="stratified",
        fractions={"train": 0.7, "validation": 0.1, "test": 0.2},
        target_col="target",
        seed=42,
        k=5,
    )
    record(
        "PolarsFE vNext",
        "model_prep",
        n,
        lambda: PolarsFE.polars_apply_partition_plan(df, PolarsFE.polars_fit_partition_plan(df, partition_plan)),
    )

    record(
        "Polars eager direct",
        "model_prep",
        n,
        lambda: df.with_row_index(".row_id", offset=1).with_columns(
            [
                pl.when(pl.arange(0, pl.len()) % 10 < 7)
                .then(pl.lit("train"))
                .when(pl.arange(0, pl.len()) % 10 == 7)
                .then(pl.lit("validation"))
                .otherwise(pl.lit("test"))
                .alias(".partition"),
                ((pl.arange(0, pl.len()) % 5) + 1).alias(".fold_id"),
            ]
        ),
    )

    record(
        "Polars eager direct",
        "numeric",
        n,
        lambda: df.with_columns(
            [
                pl.col("x").log1p().alias("x_log1p"),
                pl.col("x").sqrt().alias("x_sqrt"),
                ((pl.col("x") - pl.col("x").mean()) / pl.col("x").std()).alias("x_standardize"),
            ]
        ),
    )

    record(
        "Polars lazy direct",
        "numeric",
        n,
        lambda: df.lazy()
        .with_columns(
            [
                pl.col("x").log1p().alias("x_log1p"),
                pl.col("x").sqrt().alias("x_sqrt"),
                ((pl.col("x") - pl.col("x").mean()) / pl.col("x").std()).alias("x_standardize"),
            ]
        )
        .collect(),
    )

    try:
        import pandas as pd

        pdf = df.to_pandas()
        record(
            "pandas direct",
            "numeric",
            n,
            lambda: pdf.assign(
                x_log1p=lambda d: (d["x"] + 1).map(__import__("math").log),
                x_sqrt=lambda d: d["x"] ** 0.5,
                x_standardize=lambda d: (d["x"] - d["x"].mean()) / d["x"].std(),
            ),
        )
    except Exception as exc:  # noqa: BLE001
        rows.append({"engine": "pandas direct", "family": "numeric", "rows": n, "elapsed_seconds": 0, "status": "skipped", "error": str(exc)})

    with (OUT_DIR / "python_feature_engineering_summary.csv").open("w", newline="", encoding="utf-8") as fh:
        writer = csv.DictWriter(fh, fieldnames=["engine", "family", "rows", "elapsed_seconds", "status", "error"])
        writer.writeheader()
        writer.writerows(rows)

    decisions = {}
    for row in rows:
        if row["status"] != "success":
            continue
        family = row["family"]
        if family not in decisions or row["elapsed_seconds"] < decisions[family]["elapsed_seconds"]:
            decisions[family] = row
    with (OUT_DIR / "python_decision_table.csv").open("w", newline="", encoding="utf-8") as fh:
        writer = csv.DictWriter(fh, fieldnames=["family", "engine", "elapsed_seconds", "decision"])
        writer.writeheader()
        for family, row in decisions.items():
            writer.writerow(
                {
                    "family": family,
                    "engine": row["engine"],
                    "elapsed_seconds": row["elapsed_seconds"],
                    "decision": f"{row['engine']} was fastest in this smoke run; confirm with moderate/overnight matrix.",
                }
            )

    (OUT_DIR / "system_info_python.json").write_text(
        json.dumps({"python": sys.version, "platform": platform.platform(), "machine": platform.machine()}, indent=2),
        encoding="utf-8",
    )

    md = ["# Feature Engineering Python Smoke Benchmark", "", f"Output path: {OUT_DIR.as_posix()}", "", "## Results", ""]
    for row in rows:
        md.append(f"- {row['engine']} / {row['family']}: {row['status']} in {row['elapsed_seconds']} sec {row['error']}")
    (OUT_DIR / "feature_engineering_python_summary.md").write_text("\n".join(md), encoding="utf-8")
    print(OUT_DIR.as_posix())


if __name__ == "__main__":
    main()

File: FeatureEngineering/run_feature_engineering_smoke.R
options(stringsAsFactors = FALSE)

bench_dir <- normalizePath(file.path(getwd(), "FeatureEngineering"), mustWork = FALSE)
if (!dir.exists(bench_dir)) {
  bench_dir <- normalizePath(dirname(commandArgs(trailingOnly = FALSE)[grep("--file=", commandArgs(trailingOnly = FALSE))[1L]]), mustWork = FALSE)
}

timestamp <- format(Sys.time(), "%Y%m%d_%H%M%S")
out_dir <- Sys.getenv("FEATURE_ENGINEERING_BENCHMARK_OUTPUT", unset = file.path(bench_dir, "outputs", timestamp))
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)

rodeo_root <- normalizePath(file.path(getwd(), "..", "Rodeo"), mustWork = FALSE)
if (!dir.exists(rodeo_root)) {
  rodeo_root <- "C:/Users/Bizon/Documents/GitHub/Rodeo"
}

library(data.table)
source(file.path(rodeo_root, "R", "FeatureEngineering_vNext.R"))
source(file.path(rodeo_root, "R", "ModelPrep_vNext.R"))

make_data <- function(n = 10000L) {
  set.seed(42L)
  data.table(
    id = seq_len(n),
    x = runif(n, 0, 100),
    y = rnorm(n, 50, 10),
    cat = sample(c("A", "B", "C", "D", "E"), n, TRUE),
    cat2 = sample(c("K", "L", "M"), n, TRUE),
    target = sample(c("no", "yes"), n, TRUE),
    group = sample(paste0("g", 1:200), n, TRUE),
    date = as.Date("2024-01-01") + sample(0:365, n, TRUE),
    text = sample(c("Hello WORLD", "two words", "ABC123!", "small", "", NA_character_), n, TRUE)
  )
}

time_it <- function(engine, family, expression) {
  gc()
  elapsed <- system.time(force(expression))[["elapsed"]]
  data.table(engine = engine, family = family, rows = nrow(data), elapsed_seconds = as.numeric(elapsed), status = "success")
}

data <- make_data()
results <- list()

plan_numeric <- rodeo_feature_plan(numeric = list(columns = c("x", "y"), transforms = c("log1p", "sqrt", "standardize", "winsorize"), winsorize_probs = c(0.01, 0.99)))
results[[length(results) + 1L]] <- time_it("Rodeo vNext", "numeric", rodeo_fit_transform_feature_plan(data, plan_numeric))

results[[length(results) + 1L]] <- time_it("data.table direct", "numeric", {
  dt <- copy(data)
  set(dt, j = "x_log1p", value = log1p(dt$x))
  set(dt, j = "x_sqrt", value = sqrt(dt$x))
  set(dt, j = "x_standardize", value = (dt$x - mean(dt$x)) / sd(dt$x))
  dt
})

plan_combined <- rodeo_feature_plan(
  numeric = list(columns = c("x", "y"), transforms = c("log1p", "sqrt", "standardize", "winsorize"), winsorize_probs = c(0.01, 0.99)),
  categorical = list(columns = "cat", top_n = 4L, rare_level = "__RARE__", unseen_level = "__UNSEEN__", one_hot = TRUE, keep_original = TRUE),
  calendar = list(columns = "date", features = c("year", "month", "wday", "quarter", "is_weekend")),
  text = list(columns = "text", features = c("char_count", "word_count", "digit_count", "blank")),
  missingness = list(columns = c("x", "cat", "text"), suffix = "_is_missing"),
  interactions = list(numeric_pairs = list(c("x", "y")), categorical_numeric = list(list(categorical = "cat", numeric = "x")), categorical_pairs = list(c("cat", "cat2")), max_features = 20L)
)
results[[length(results) + 1L]] <- time_it("Rodeo vNext", "combined_plan", rodeo_fit_transform_feature_plan(data, plan_combined))

partition_plan <- rodeo_partition_plan(
  method = "stratified",
  fractions = c(train = 0.7, validation = 0.1, test = 0.2),
  target_col = "target",
  seed = 42L,
  k = 5L
)
results[[length(results) + 1L]] <- time_it("Rodeo vNext", "model_prep", {
  fitted <- rodeo_fit_partition_plan(data, partition_plan)
  rodeo_apply_partition_plan(data, fitted)
})

results[[length(results) + 1L]] <- time_it("data.table direct", "model_prep", {
  dt <- copy(data)
  set(dt, j = ".row_id", value = seq_len(nrow(dt)))
  dt[, .partition := sample(c("train", "validation", "test"), .N, TRUE, prob = c(0.7, 0.1, 0.2)), by = target]
  dt[, .fold_id := ((seq_len(.N) - 1L) %% 5L) + 1L]
  dt
})

r_results <- rbindlist(results, use.names = TRUE)
fwrite(r_results, file.path(out_dir, "r_feature_engineering_summary.csv"))

decision <- r_results[, .SD[which.min(elapsed_seconds)], by = family]
decision[, decision := paste(engine, "was fastest in this smoke run; confirm with moderate/overnight matrix before package thresholds.")]
fwrite(decision, file.path(out_dir, "r_decision_table.csv"))

writeLines(capture.output(sessionInfo()), file.path(out_dir, "session_info.txt"))
writeLines(c(
  paste("timestamp:", timestamp),
  paste("machine:", Sys.info()[["nodename"]]),
  paste("system:", paste(Sys.info(), collapse = " | "))
), file.path(out_dir, "system_info.txt"))

md <- c(
  "# Feature Engineering R Smoke Benchmark",
  "",
  paste("Output path:", normalizePath(out_dir, winslash = "/")),
  "",
  "## Results",
  "",
  paste(capture.output(print(r_results)), collapse = "\n"),
  "",
  "## Provisional Decision Table",
  "",
  paste(capture.output(print(decision)), collapse = "\n")
)
writeLines(md, file.path(out_dir, "feature_engineering_r_summary.md"))
cat(normalizePath(out_dir, winslash = "/"), "\n")
